File: hi-none.c
#!/usr/bin/env bash
# hi-none.c - Always recompile (no caching)
FILE="${0}"
CC_FLAGS=(-std=gnu99 -O0 -pipe -I"$(dirname "$(readlink -f "${FILE}")")" -Wall -Wextra -Wno-unused)
TEMP_C="$(mktemp "${TMPDIR-/tmp}/XXX.c")"
OUT="${TEMP_C%.c}"
trap 'rm -f -- "${TEMP_C}" "${OUT}"' EXIT
sed -e "1s|^#!.*|#line 2 \"${FILE}\"|" -e '2,/^#!/ s|.*||' < "${FILE}" > "${TEMP_C}"

C_LINE=$( sed -n '1n; /^#!/{=;q}' "${FILE}" )
bat --language=C --number --paging=never --line-range="$(( C_LINE + 1 )):" "${TEMP_C}" 1>&2 2>/dev/null ||
cat -n "${TEMP_C}" | tail -n +$(( C_LINE + 1 )) 1>&2 2>/dev/null

set -x
"${CC-cc}" "${CC_FLAGS[@]}" -o"${OUT}" "${TEMP_C}" 1>&2
(exec -a "${FILE}" "${OUT}" "${@}")
exit $?
#!/usr/bin/env -S ccraft # -----------------------------------------------------
#include <stdio.h>

int
main(int argc, char *argv[]) {
    printf("hello %s!\n", argc > 1 ? argv[1] : "world");
    return 0;
}

File: hi-none.go
#!/usr/bin/env bash
# hi-none.go — Go script runner, always rebuild
FILE=$(readlink -f "${0}") || exit 1
BASE="${FILE##*/}"
DIR=$(mktemp -d "/tmp/${BASE%.go}-XXXXX")
OUT=$(mktemp "/tmp/${BASE%.go}-XXXXX")
trap 'rm -rf "${DIR}" "${OUT}"' EXIT

sed -e '1s|#!|//|' -e '2,/^#!/ s|.*||' < "${FILE}" > "${DIR}/main.go"

CLINE=$(sed -n '1n; /^#!/{=;q}' "${FILE}")
bat --language=go --number --paging=never --line-range="$(( CLINE + 1 )):" "${DIR}/main.go" 2>/dev/null ||
cat -n "${DIR}/main.go" | tail -n +"$(( CLINE + 1 ))" 2>/dev/null

cd "${DIR}"
set -x
"${GO-go}" mod init "${BASE%.go}" 2>/dev/null
"${GO-go}" mod tidy 2>&1
"${GO-go}" build -buildvcs=false -o "${OUT}" . 1>&2

(exec -a "${FILE}" "${OUT}" "${@}")
exit $?
#!/usr/bin/env -S goon
package main

import (
    "fmt"
    "rsc.io/quote"
    "os"
)

func main() {
    arg := "world"
    if len(os.Args) > 1 {
        arg = os.Args[1]
    }
    fmt.Printf("hello %s!\n%s\n", arg, quote.Hello())
}
